cli/azd/cmd/templates.go

```diff
@@ -175,3 +208,290 @@ func getCmdTemplateHelpDescription(*cobra.Command) string {
+	return &templateSourceAddAction{
+		flags:         flags,
+		console:       console,
+		sourceManager: sourceManager,
+		args:          args,
+	}
+}
+
+func (a *templateSourceAddAction) Run(ctx context.Context) (*actions.ActionResult, error) {
+	a.console.MessageUxItem(ctx, &ux.MessageTitle{
+		Title: "Add azd template source (azd template source add)",
+	})
+
+	var key = a.args[0]
+	sourceConfig := &templates.SourceConfig{}
+
+	if _, ok := templates.WellKnownSources[key]; !ok {
+		sourceConfig = &templates.SourceConfig{
+			Key:      key,
+			Type:     templates.SourceKind(a.flags.kind),
+			Location: a.flags.location,
+			Name:     a.flags.name,
+		}
+
+		// Validate the custom source config
+		spinnerMessage := "Validating template source"
+		a.console.ShowSpinner(ctx, spinnerMessage, input.Step)
+		_, err := a.sourceManager.CreateSource(ctx, sourceConfig)
+		if err != nil {
+			a.console.StopSpinner(ctx, spinnerMessage, input.StepFailed)
+			return nil, fmt.Errorf("template source validation failed: %w", err)
+		}
+
+		a.console.StopSpinner(ctx, spinnerMessage, input.StepDone)
+	}
+
+	spinnerMessage := "Saving template source"
+	a.console.ShowSpinner(ctx, spinnerMessage, input.Step)
+	if err := a.sourceManager.Add(ctx, key, sourceConfig); err != nil {
+		a.console.StopSpinner(ctx, spinnerMessage, input.StepFailed)
+		return nil, fmt.Errorf("failed adding template source: %w", err)
+	}
+
+	a.console.StopSpinner(ctx, spinnerMessage, input.StepDone)
+
+	return &actions.ActionResult{
+		Message: &actions.ResultMessage{
+			Header:   fmt.Sprintf("Added azd template source %s", key),
+			FollowUp: "Run `azd template list` to see the available set of azd templates.",
+		},
+	}, nil
+}
+
+func newTemplateSourceRemoveCmd() *cobra.Command {
+	return &cobra.Command{
+		Use:   "remove <key>",
+		Short: fmt.Sprintf("Removes the specified azd template source %s", output.WithWarningFormat("(Beta)")),
+		Args:  cobra.ExactArgs(1),
+	}
+}
+
+type templateSourceRemoveAction struct {
+	sourceManager templates.SourceManager
+	console       input.Console
+	args          []string
+}
+
+func newTemplateSourceRemoveAction(
+	sourceManager templates.SourceManager,
+	console input.Console,
+	args []string,
+) actions.Action {
+	return &templateSourceRemoveAction{
+		sourceManager: sourceManager,
+		console:       console,
+		args:          args,
+	}
+}
+
+func (a *templateSourceRemoveAction) Run(ctx context.Context) (*actions.ActionResult, error) {
+	a.console.MessageUxItem(ctx, &ux.MessageTitle{
+		Title: "Remove azd template source (azd template source remove)",
+	})
+
+	var key = a.args[0]
+	spinnerMessage := "Removing template source"
+	a.console.ShowSpinner(ctx, spinnerMessage, input.Step)
+	if err := a.sourceManager.Remove(ctx, key); err != nil {
+		a.console.StopSpinner(ctx, spinnerMessage, input.StepFailed)
+		return nil, fmt.Errorf("failed removing template source: %w", err)
+	}
+
+	a.console.StopSpinner(ctx, spinnerMessage, input.StepDone)
```

Comment:
Same as previous comment. Consider using `input.GetStepResultFormat()`.

Or, you can also consider using `defer` for StopSpinner() like we here:

![image](https://github.com/Azure/azure-dev/assets/24213737/833bf1a8-eb9b-40a9-9954-abd12c146385)